DOCS 868: Document the RoboClaw motor (#1453)
viamrobotics/docs: docs/components/motor/roboclaw.md

---
title: "Configure a roboclaw motor"
linkTitle: "roboclaw"
weight: 22
type: "docs"
description: "Configure a brushed DC motor driven by a roboclaw motor controller."
images: ["/components/img/components/motor.svg"]
# SMEs: Olivia, Rand
---

The `roboclaw` model of the motor component supports [standard brushed DC motors](https://en.wikipedia.org/wiki/DC_motor) driven by [Basicmicro's](https://www.basicmicro.com/) [RoboClaw](https://www.basicmicro.com/RoboClaw-2x30A-Motor-Controller_p_9.html) motor controller.

{{< alert title="Note" color="note" >}}

> You must set up your RoboClaw before configuring it.

Reviewer: [nit] Is this order of operations necessary because you wouldn't know the motor channel or something until after you set it up?

> You must set up your RoboClaw before configuring it.

Reviewer 2: ```suggestion
You must set up your RoboClaw before configuring it. Make note of the baud rate that you set up during this process, the default is 38400.
```

Follow [this guide](https://resources.basicmicro.com/roboclaw-motor-controllers-getting-started-guide/) to do so.

{{< /alert >}}

To configure a `roboclaw` motor as a component of your robot:

{{< tabs name="Configure your roboclaw motor">}}
{{% tab name="Config Builder" %}}

Navigate to the **Config** tab of your robot's page in [the Viam app](https://app.viam.com).
Click on the **Components** subtab and navigate to the **Create component** menu.
Enter a name for your motor, select the type `motor`, and select the `roboclaw` model.

Click **Create component**.

![A roboclaw motor config.](../../img/motor/roboclaw-ui-config.png)

Edit and fill in the attributes as applicable.

{{% /tab %}}
{{% tab name="JSON Template" %}}

```json
{
  "components": [
    {

>       "name": <"your-roboclaw-motor-name">,

Reviewer: ```suggestion
      "name": "<your-roboclaw-motor-name>",
```

      "type": "motor",
      "model": "roboclaw",
      "attributes": {

>         "serial_path": <"your-serial-path">,

Reviewer: ```suggestion
        "serial_path": "<your-serial-path>",
```

>         "motor_channel": <#>,
>         "serial_baud_rate": <#>

Reviewer: ```suggestion
        "motor_channel": <int>,
        "serial_baud_rate": <int>
```
[opt nit] It seems that this is quite a bit more common in the docs right now

Author: Ah yes, good catch!

      },
      "depends_on": []
    }
  ]
}
```

{{% /tab %}}
{{% tab name="JSON Example" %}}

Example configuration for a `roboclaw` DC brushed motor:

```json
{
  "components": [
    {
      "name": "your-roboclaw-motor",
      "type": "motor",
      "model": "roboclaw",
      "attributes": {
        "serial_path": "/dev/ttyUSB0",
        "motor_channel": 1,
        "serial_baud_rate": 38400
      },
      "depends_on": []
    }
  ]
}
```

{{% /tab %}}
{{< /tabs >}}

The following attributes are available for `roboclaw` motors:

| Name | Type | Inclusion | Description |
| ---- | ---- | --------- | ----------- |

> | `serial_path` | string | **Required** | Serial path of the `roboclaw` controller's USB connection to your robot's filesystem. Run `ls /dev/tty*` in your terminal to find USB serial ports on Linux. <br> Example: `"/dev/ttyUSB0"` |

Reviewer 3: When I was using the roboclaw on my pi my path was `/dev/serial/by-path/...` so this line might cause confusion.

Author: @oliviamiller Ah great to know and change here thank you!

> | `serial_path` | string | **Required** | Serial path of the `roboclaw` controller's USB connection to your robot's filesystem. Run `ls /dev/tty*` in your terminal to find USB serial ports on Linux. <br> Example: `"/dev/ttyUSB0"` |

Reviewer 3: Yeah that looks good! my full path was `/dev/serial/by-path/platform-fd500000.pcie-pci-0000:01:00.0-usb-0:1.1:1.0`

> | `serial_baud_rate` | int | **Required** | [Rate to send data](https://learn.sparkfun.com/tutorials/serial-communication) over the serial line. You cannot have multiple `roboclaw` motors with different baud rates. <br> Default: `38400` |

Reviewer: If this is required, how is there a default? Not seeing it default to anything in https://github.com/viamrobotics/rdk/blob/main/components/motor/roboclaw/roboclaw.go

Reviewer 3: The user must specify a baud rate in the config but the default when setting up the roboclaw on their motion studio application is 38400 - so might be worth noting that in the docs.

Reviewer 2: We can remove the requirement, but whatever you set up using their setup program is what the baudrate becomes, added a few sentences to clarify.

Author: @oliviamiller @randhid Cool, thank you!

> | `serial_baud_rate` | int | **Required** | [Rate to send data](https://learn.sparkfun.com/tutorials/serial-communication) over the serial line. You cannot have multiple `roboclaw` motors with different baud rates. <br> Default: `38400` |

Reviewer 2: ```suggestion
| `serial_baud_rate` | int | Optional | [Rate to send data](https://learn.sparkfun.com/tutorials/serial-communication) over the serial line. This must match the baudrate you have set up using basicmicro's setup program. You cannot have multiple `roboclaw` motors with different baud rates. <br> Default: `38400` |
```

Author: Done, ty!
